src/client/core/page-unload-barrier.js

```diff
@@ -0,0 +1,130 @@
+import hammerhead from './deps/hammerhead';
+import * as eventUtils from './utils/event';
+import { EventEmitter } from './utils/service';
+import delay from './utils/delay';
+
+var Promise      = hammerhead.Promise;
+var browserUtils = hammerhead.utils.browser;
+
+
+const DEFAULT_BARRIER_TIMEOUT       = 500;
+const WAIT_FOR_UNLOAD_TIMEOUT       = 3000;
+const SHORT_WAIT_FOR_UNLOAD_TIMEOUT = 30;
+
+const WAITING_FINISHED_EVENT = 'waiting-finished';
+
+
+class PageUnloadMonitor extends EventEmitter {
+    constructor () {
+        super();
+
+        this.waitingForUnload          = false;
+        this.waitingForUnloadTimeoutId = null;
+        this.unloading                 = false;
+
+        this._handleSubmit();
+        this._handleLinkClick();
+        this._handleBeforeUnload();
+    }
+
+    _overrideFormSubmit (form) {
+        var submit = form.submit;
+
+        form.submit = () => {
+            this._prolongUnloadWaiting(WAIT_FOR_UNLOAD_TIMEOUT);
+            submit.apply(form, arguments);
+        };
+    }
+
+    _handleSubmit () {
+        eventUtils.bind(document, 'submit', e => {
+            if (e.target.tagName.toLowerCase() === 'form')
+                this._prolongUnloadWaiting(WAIT_FOR_UNLOAD_TIMEOUT);
+        });
+
+        var forms = document.getElementsByTagName('form');
+
+        for (var i = 0; i < forms.length; i++)
+            this._overrideFormSubmit(forms[i]);
+    }
+
+    _handleLinkClick () {
+        eventUtils.bind(document, 'click', e => {
+            var target = e.srcElement || e.target;
+
+            if (!e.defaultPrevented && target.tagName && target.tagName.toLowerCase() === 'a') {
+                var href = target.href;
+
+                if (target.hasAttribute('href') && !/(^javascript:)|(^mailto:)|(^tel:)|(^#)/.test(href))
+                    this._prolongUnloadWaiting(WAIT_FOR_UNLOAD_TIMEOUT);
+            }
+        });
+    }
+
+    _onBeforeUnload () {
+        if (!browserUtils.isIE) {
+            this.unloading = true;
+            return;
+        }
+
+        this._prolongUnloadWaiting(SHORT_WAIT_FOR_UNLOAD_TIMEOUT);
+
+        delay(0)
+            .then(() => {
+                //NOTE: except file downloading
+                if (document.readyState === 'loading' &&
+                    !(document.activeElement && document.activeElement.tagName.toLowerCase() === 'a' &&
```

Comment:
There is no such function in [`hammerhead.domUtils`](https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/utils/dom.js)